### ISSUE TITLE
Bug: defaultOptions does not provide default options

### ISSUE THREAD

**Issue by gajus** [typescript-eslint/typescript-eslint#6438]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
N/A
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

If I provide `defaultOptions`, then I expect to get those from `context.options` if no other options are provided.

### Actual Result

Despite providing `defaultOptions`, `options.context` comes back as an empty array.

### Versions

N/A

**Comment by Josh-Cena** [comment]
This is working as intended. You should use the `options` parameter instead. See https://github.com/typescript-eslint/typescript-eslint/issues/6072

**Comment by gajus** [comment]
> This is working as intended. You should use the `options` parameter instead. See #6072

Thank you. What is `defaultOptions` then and where is this documented? Cannot find any references

**Comment by gajus** [comment]
> > This is working as intended. You should use the `options` parameter instead. See #6072
> 
> Thank you. What is `defaultOptions` then and where is this documented? Cannot find any references

Never mind. Misread your response. I now get it, but is this documented anywhere?

**Comment by Josh-Cena** [comment]
I don't think so, tbh—in fact `RuleCreator` may not even be part of the public documented API. @JoshuaKGoldberg @bradzacher does it seem like something worth mentioning somewhere?

**Comment by gajus** [comment]
It is documented here https://typescript-eslint.io/custom-rules

**Comment by Josh-Cena** [comment]
Oh, fair enough. Interesting that the docs don't even mention the `options` parameter.

**Comment by bradzacher** [comment]
The reason we don't modify `context.options` is because (a) ESLint freezes the `context` object plus ESLint uses prototypal inheritance for the `context` - so we'd need to clone it to modify it and (b) if we modify it there would be no way to get the user's original options, in case that's required.

Instead yes, we pass a second argument to the `create` call which is the options with the defaults applied.

It's not documented because our docs are quite incomplete! Happy to have it added though!

**Comment by gajus** [comment]
A related issue, I don't know if documentation or bug, but it seems like you merge default values using `deepMerge`, but types do not take into account the default values, i.e. I have

```ts
type Options = [
  {
    matchCallExpression?: boolean;
    suffix?: string | null;
    transforms?: string[] | string | null;
  },
];
```

and

```ts
defaultOptions: [
    {
      matchCallExpression: false,
      suffix: null,
      transforms: null,
    },
  ],
```

given these two values, and the fact that you use deep merge, I would expect that types for options say that all of the object properties will be always present, whereas now they are all marked as optional.

**Comment by bradzacher** [comment]
The problem is that TS doesn't really let you do everything you want to to make it work.

First problem is that predicting how TS infers generics can be impossible sometimes - in some cases TS could infer the correct set of keys for `TMessageIds`, and other times it may not because there is a usage of `context.report` that passes a new message ID, so it widens to `string`. It's just hard to get right - so we kind of want users to pass in an explicit type so that you don't rely on TS inference if possible for it.

The second problem is that there's currently no way to tell TS "please use this type for this generic, but infer this other generic". Which is an issue because there's no way for us to build any tooling that can infer the options type from a JSON schema object without building a huge amount of infra around it.
So we *need* consumers to pass in a `TOptions` parameter, and because of that we also need users to pass in a `TMessageIds` parameter.

Now the default options array is technically a separate, third generic in the function. Now - again - we cannot have TS infer the type of the defaults based on the usage whilst also providing explicit types for the others. So if you want to provide the functionality to "merge" the default option type with the base option type to provide a correct type for the default options, then you would need to require users to pass a third type parameter - meaning they'd need to explicitly type their default options.

It's just more config work which, whilst nice, is hard because (a) lot of config objects can get pretty complicated with defaults and (b) generically merging two unknown types together isn't a super easy task in of itself.

It's something I wanted when I build the tooling but ditched early on because it wasn't really worth the time to build or the effort from users to type, IMO.

**Comment by gajus** [comment]
This makes sense. I think similar to the earlier issue, this could be simply solved with documentation, e.g. by linking to this issue and providing an example strategy for dealing with this.
